File: algorithm-design/hackerrank/the_grid_search/the_grid_search.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/the-grid-search

import os
import unittest

from typing import List


def grid_search(G: List[str], P: List[str]) -> str:
    everything = ''.join(G)
    at = -1
    while True:
        at = everything.find(P[0], at + 1)
        if at == -1:
            return 'NO'
        if (at % len(G[0])) + len(P[0]) > len(G[0]):
            continue
        offset = at
        ok = True
        for chunk in P:
            if everything[offset:offset + len(chunk)] != chunk:
                ok = False
                break
            offset += len(G[0])
        if ok:
            return 'YES'


class TestCode(unittest.TestCase):
    def runner(self, name):
        path = os.path.join(os.path.split(__file__)[0], f'input{name}.txt')
        with open(path, 'r') as handle:
            lines = handle.readlines()
            lines = [line.strip() for line in lines]
        tests = int(lines[0])
        results = [None] * tests
        offset = 1
        for test in range(tests):
            r_g = int(lines[offset].split(' ')[0])
            g = lines[offset + 1:offset + 1 + r_g]
            r_p = int(lines[offset + 1 + r_g].split(' ')[0])
            p = lines[offset + 1 + r_g + 1:offset + 1 + r_g + 1 + r_p]
            offset += r_g + r_p + 2
            results[test] = grid_search(g, p)
        path = os.path.join(os.path.split(__file__)[0], f'output{name}.txt')
        with open(path, 'r') as handle:
            lines = handle.readlines()
            expected = [line.strip() for line in lines]
        self.assertEqual(expected, results)

    def test_example(self):
        self.runner('_example')

    def test_05(self):
        self.runner('05')

    def test_07(self):
        self.runner('07')

    def test_08(self):
        self.runner('08')

    def test_09(self):
        self.runner('09')

    def test_15(self):
        self.runner('15')

File: algorithms/code/hackerrank/the_grid_search/the_grid_search.rs
// https://www.hackerrank.com/challenges/the-grid-search

use crate::hackerrank::tester::{read_input, write_and_check_output};

pub fn grid_search(g: &[String], p: &[String]) -> String {
    let everything = g.join("");
    let mut at: usize = 0;
    loop {
        if let Some(result) = everything[at..].find(&p[0]) {
            at += result;
        } else {
            return "NO".to_string();
        }
        if at % g[0].len() + p[0].len() > g[0].len() {
            at += 1;
            continue;
        }
        let mut offset = at;
        let mut ok = true;
        for chunk in p {
            if everything[offset..offset + chunk.len()] != *chunk {
                ok = false;
                break;
            }
            offset += g[0].len();
        }
        if ok {
            return "YES".to_string();
        }
        at += 1;
    }
}

pub fn main_tested(name: &str) {
    let origin = file!();
    let input = read_input(origin, name);
    let t = input[0][0].parse::<i32>().unwrap();
    let mut results: Vec<String> = Vec::new();
    let mut offset = 1;
    for _ in 0..t {
        let g_length = input[offset][0].parse::<usize>().unwrap();
        let g: Vec<String> = input[offset + 1..offset + 1 + g_length]
            .iter()
            .map(|line| line[0].clone())
            .collect();
        let p_length = input[offset + 1 + g_length][0].parse::<usize>().unwrap();
        let p: Vec<String> = input[offset + 1 + g_length + 1..offset + 1 + g_length + 1 + p_length]
            .iter()
            .map(|line| line[0].clone())
            .collect();
        offset += 1 + g_length + 1 + p_length;
        results.push(grid_search(&g, &p));
    }
    write_and_check_output(origin, name, &results);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        main_tested("_example");
    }

    #[test]
    fn test05() {
        main_tested("05");
    }

    #[test]
    fn test07() {
        main_tested("07");
    }

    #[test]
    fn test08() {
        main_tested("08");
    }

    #[test]
    fn test09() {
        main_tested("09");
    }

    #[test]
    fn test15() {
        main_tested("15");
    }
}
